streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImpl.java

```diff
@@ -924,7 +925,7 @@ public void process(final ProcessorSupplier<? super K, ? super V> processorSuppl
             allSourceNodes.addAll(((KStreamImpl<K1, V2>) other).sourceNodes);
 
             // do not have serde for joined result
-            return new KStreamImpl<>(joinMergeName, joined.keySerde(), null, allSourceNodes, false, joinGraphNode, builder);
+            return new KStreamImpl<K1, R>(joinMergeName, joined.keySerde() != null ? joined.keySerde() : ((AbstractStream) lhs).keySerde(), null, allSourceNodes, false, joinGraphNode, builder);
```

Comment:
This is found and fixed while I'm adding unit tests.

Reply:
This is similar to `branch()` -- we arbitrarily pick left hand side input. Also, if lhs is `null` rhs could provide correct serde. \cc @vvcephei Can you add this case to the JIRA for branch(), too.

Reply:
Ack.